packages/yarnpkg-pnpify/sources/generateSdk.ts

```diff
@@ -22,7 +22,7 @@ const TEMPLATE = (relPnpApiPath: PortablePath, module: string, {setupEnv = false
   `const absPnpApiPath = resolve(__dirname, relPnpApiPath);\n`,
   `const absRequire = (createRequire || createRequireFromPath)(absPnpApiPath);\n`,
   `\n`,
-  `if (existsSync(absPnpApiPath)) {\n`,
+  `if (existsSync(absPnpApiPath) && !process.versions.pnp) {\n`,
```

Comment:
It won't work for tools using different `setupEnv` / `usePnpify` settings (for example, if the tool without setupEnv executes first, then the env will never be setup). I think we'll need to use non-enumerable global variables to support this:

```
if (typeof __yarnpkg_sdk_has_setup_env__ === `undefined`) {
  Object.defineProperty(global, `__yarnpkg_sdk_has_setup_env__`, {
    enumerable: false,
    value: true,
  });

  process.env.NODE_OPTIONS = ...;
}
```

Reply:
hmm, I couldn't find the code path that set setupEnv or usePnpify to true. Who's actually use those 2 flags and what's the intention for having that?

Reply:
Some tools used to spawn child processes and needed help forwarding the hook, and the old TS layer used to require the pnpify layer. I think both extensions that required that got changed upstream, so we don't need them anymore at the moment.

That said, we should either remove this codepath altogether, or keep supporting it even through the singleton - I would probably opt for the second one (keeping them) because I wouldn't be surprised if we were to need them in a pinch.

Reply:
> for example, if the tool without setupEnv executes first, then the env will never be setup
I still have difficulties understand this use cases, would that be possible to give me a code snippet to illustrate this?

Reply:
Forgot to comment back! Basically, if you instantiate an extension with `usePnpify=false`, then another with `usePnpify=true`, PnPify wouldn't have been loaded at all because the first one would have set `process.versions.pnp`, and the second one would have seen that and bailout. By using three differen value, we can be sure that the different patches are applied in isolation.